cmd/sampleVcf/sampleVcf.go

```diff
@@ -0,0 +1,100 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"github.com/vertgenlab/gonomics/common"
+	"github.com/vertgenlab/gonomics/fileio"
+	"github.com/vertgenlab/gonomics/vcf"
+	"log"
+	"math/rand"
+)
+
+//sampleVcf takes a VCF file and returns a random subset of variants to an output VCF file. Can also retain a random subset of alleles from gVCF data (diploid, does not break allele pairs)
+func sampleVcf(inFile string, outFile string, numVariants int, numSamples int, randSeed bool, setSeed int64) {
+	common.RngSeed(randSeed, setSeed)
+	records, header := vcf.ReadWithHeader(inFile)
+
+	var sampleList []string = vcf.HeaderGetSampleList(header)
+
+	if numVariants > len(records) {
+		log.Fatalf("The number of requested sampled variants is greater than the number of variants in the input file.")
+	}
+
+	//Shuffle the vcf records, our subset will be composed to the first entries in the shuffled order.
+	rand.Shuffle(len(records), func(i, j int) { records[i], records[j] = records[j], records[i] })
+	//DEBUG:fmt.Printf("lenRecords before slice: %v.\n", len(records))
+	records = records[:numVariants] //keep only as many results as specified.
+	//DEBUG: fmt.Printf("lenRecords after slice: %v.\n", len(records))
+
+	if numSamples > 0 {
+		if numSamples > len(records[0].Samples) {
+			log.Fatalf("More samples were requested than were present in the input VCF file.")
+		}
+		var sequentialSlice []int = getSequentialSlice(len(records[0].Samples))
+		rand.Shuffle(len(sequentialSlice), func(i, j int) { sequentialSlice[i], sequentialSlice[j] = sequentialSlice[j], sequentialSlice[i] })
+		sequentialSlice = sequentialSlice[:numSamples] //now we have a list of samples to keep from each variant.
+
+		var outHeaderSampleList []string = make([]string, 0)
+		for _, i := range sequentialSlice {
+			outHeaderSampleList = append(outHeaderSampleList, sampleList[i])
+		}
+
+		vcf.HeaderUpdateSampleList(header, outHeaderSampleList)
+
+		var outSamples []vcf.GenomeSample
+
+		for _, i := range records {
+			outSamples = make([]vcf.GenomeSample, 0)
```

Comment:
Minor point, if you are sampling a large number of variants than this will slow you down a lot as you need to keep reshuffling memory as you append elements beyond the capacity of this slice. You could speed this up by doing `outSamples = make([]vcf.GenomeSample, 0, len(sequentialSlice)`. Giving it the cap will make append much faster (as i understand it it would be equivalent to doing outSamples[i] = i.Samples[j]). Better yet I think you could even declare outSamples outside of the loop and just reuse the memory each run.

That said, I suspect this is not a performance-intensive task so I will approve anyways.